### PR TITLE
Stateless ladspa effects

### PR DESCRIPTION
**Pull request by Paul-Licameli**
Resolves: #2768

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

File: src/effects/ladspa/LadspaEffect.cpp [hunk 1]
```diff
@@ -111,6 +111,15 @@ LadspaEffectsModule::~LadspaEffectsModule()
 {
 }
 
+// Don't use the template-generated MakeSettings(), which default-constructs
+// the structure.  Instead allocate a number of values chosen by thge plug-in
```
[the remark this answers is not in the document]
Comment (Paul-Licameli) [reply src/effects/ladspa/LadspaEffect.cpp:115]: Maybe I was trying to spell this word https://www.thefreedictionary.com/apophthegm

**Comment by pietro68** [comment]
I think I found a regression: when using a ladspa effect in realtime mode, changes in the sliders positions are ignored.

Try for instance with the AM pitch shifter. Generate a sine tone, then try moving the pitch shift slider left and right while the tone is being played (or use the "start playback" button) - with the latest release I can hear the pitch being changed, but not with the build made from this branch.

**Comment by Paul-Licameli** [comment]
> I think I found a regression: when using a ladspa effect in realtime mode, changes in the sliders positions are ignored.
> 
> Try for instance the AM pitch shifter. Generate a sine tone, instantiate an AM pitch shifter, then try moving the pitch shift slider left and right while the tone is being played (or use the "start playback" button) - with the latest release I can hear the pitch being changed, but not with the build made from this branch.

Could you perhaps bisect the problem within this branch?

**Comment by Paul-Licameli** [comment]
> > I think I found a regression: when using a ladspa effect in realtime mode, changes in the sliders positions are ignored.
> > Try for instance the AM pitch shifter. Generate a sine tone, instantiate an AM pitch shifter, then try moving the pitch shift slider left and right while the tone is being played (or use the "start playback" button) - with the latest release I can hear the pitch being changed, but not with the build made from this branch.
> 
> Could you perhaps bisect the problem within this branch?

Never mind doing that.

I looked at this problem and I see that:

- LadspaEffect::CopySettingsContents is never reached.  That is where communication of slider changes is supposed to happen.
- That is supposed to happen in the line `mpAccessState->WorkerRead();` in RealtimeEffectState.cpp
- Which is called during contruction of a ProcessingScope in AudioIO::TransformPlayBuffers in the buffer-exchange thread.
- But that line is skipped because `mwAccess.expired()` is true
- And mwAccess becomes non-null only in the function RealtimeEffectState::GetAccess()
- And that function is not yet called anywhere.

So:  you recall reviewing my changes in RealtimeEffectState, when I introduced GetAccess() and the associated classes that are supposed to communicate setting changes across threads.  But never calling GetAccess() anywhere is an incompleteness still in our new framework.

So there's not really an error (no evidence of one anyway) special to this pull request for Ladspa effects, which is written to trust the new framework.

So the fix should be figured out and pushed as a fix before this branch is rebased onto that and pushed.

**Comment by Paul-Licameli** [comment]
Issue #2849 is now open for this problem of applying the effect in realtime.

Let that problem be fixed later.  For now, test Ladspa effects only destructively, with the Apply button.  Don't let this other problem block review and merge of this PR.

**Review by pietro68 — APPROVED** [review]
I understand the changes and I tested successfully about 6 different LADSPA effects in destructive mode.

**Comment by Paul-Licameli** [comment]
Convert LADSPA effects to new stateless system -- known problem: ...

... The effect does not respond to adjustment of sliders does not work during realtime playback.

The fault lies in the framework for stateless effects, not an error in these commits, themselves.

The cause is known and will soon be fixed.

**Pull request merged**